core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -409,20 +411,31 @@ TranslogWriter createWriter(long fileGeneration) throws IOException {
     public Location add(final Operation operation) throws IOException {
         final ReleasableBytesStreamOutput out = new ReleasableBytesStreamOutput(bigArrays);
         try {
-            final BufferedChecksumStreamOutput checksumStreamOutput = new BufferedChecksumStreamOutput(out);
             final long start = out.position();
             out.skip(Integer.BYTES);
-            writeOperationNoSize(checksumStreamOutput, operation);
+            writeOperationNoSize(new BufferedChecksumStreamOutput(out), operation);
             final long end = out.position();
             final int operationSize = (int) (end - Integer.BYTES - start);
             out.seek(start);
             out.writeInt(operationSize);
             out.seek(end);
             final ReleasablePagedBytesReference bytes = out.bytes();
+            final Location location;
             try (ReleasableLock ignored = readLock.acquire()) {
                 ensureOpen();
-                return current.add(bytes, operation.seqNo());
+                location = current.add(bytes, operation.seqNo());
             }
+            try (ReleasableLock ignored = writeLock.acquire()) {
+                if (shouldFoldGeneration(this) && foldingGeneration.compareAndSet(false, true)) {
+                    // we have to check the condition again lest we could fold twice in a race
```

Comment:
My fingers did not do what they suppose to do here. :wink: I pushed 152cae7e500a4df56f6dea9f3d835313ce972d67.